File: src/providers/native/elaboration/type-model.ts
import type { RustNativeDefinitionId } from "./evidence.js";

export type RustNativeArgument =
  | { readonly kind: "type" | "constant"; readonly id: number }
  | { readonly kind: "lifetime"; readonly region: RustNativeRegion };

export type RustNativeBoundIndex = { readonly kind: "bound"; readonly depth: number } | { readonly kind: "canonical" };
export type RustNativeBoundType =
  | { readonly kind: "anonymous" }
  | { readonly kind: "named"; readonly definition: RustNativeDefinitionId };
export type RustNativeBoundRegion =
  | { readonly kind: "anonymous" | "closure-environment" }
  | { readonly kind: "printed"; readonly name: string }
  | { readonly kind: "named"; readonly definition: RustNativeDefinitionId };
export type RustNativeLateRegion =
  | { readonly kind: "anonymous"; readonly index: number }
  | { readonly kind: "printed"; readonly index: number; readonly name: string }
  | { readonly kind: "named"; readonly definition: RustNativeDefinitionId }
  | { readonly kind: "closure-environment" };
export type RustNativeVariable =
  | { readonly kind: "type"; readonly declaration: RustNativeBoundType }
  | { readonly kind: "lifetime"; readonly declaration: RustNativeBoundRegion }
  | { readonly kind: "constant" };
export interface RustNativeBinder<Value> {
  readonly variables: readonly RustNativeVariable[];
  readonly value: Value;
}
export type RustNativeRegion =
  | { readonly kind: "early"; readonly index: number; readonly name: string }
  | { readonly kind: "bound"; readonly binder: RustNativeBoundIndex; readonly variable: number; readonly declaration: RustNativeBoundRegion }
  | { readonly kind: "late"; readonly scope: RustNativeDefinitionId; readonly declaration: RustNativeLateRegion }
  | { readonly kind: "static" | "erased" }
  | { readonly kind: "inference"; readonly index: number }
  | { readonly kind: "placeholder"; readonly universe: number; readonly variable: number; readonly declaration: RustNativeBoundRegion };

export type RustNativeAlias =
  | { readonly sort: "type"; readonly category: "projection" | "inherent" | "opaque" | "free";
      readonly definition: RustNativeDefinitionId; readonly arguments: readonly RustNativeArgument[] }
  | { readonly sort: "constant"; readonly category: "projection" | "inherent" | "free" | "anonymous";
      readonly definition: RustNativeDefinitionId; readonly arguments: readonly RustNativeArgument[] };
export interface RustNativeSignature {
  readonly inputs: readonly number[];
  readonly output: number;
  readonly variadic: boolean;
  readonly unsafeCall: boolean;
  readonly abi: string;
}
export type RustNativeExistential =
  | { readonly kind: "trait"; readonly definition: RustNativeDefinitionId; readonly arguments: readonly RustNativeArgument[] }
  | { readonly kind: "projection"; readonly definition: RustNativeDefinitionId; readonly arguments: readonly RustNativeArgument[]; readonly term: RustNativeArgument }
  | { readonly kind: "auto-trait"; readonly definition: RustNativeDefinitionId };
export type RustNativeTypePattern =
  | { readonly kind: "range"; readonly start: number; readonly end: number }
  | { readonly kind: "or"; readonly patterns: readonly RustNativeTypePattern[] }
  | { readonly kind: "not-null" };

export type RustNativeType =
  | { readonly kind: "primitive"; readonly name: "bool" | "char" | "str" | "never" |
      "i8" | "i16" | "i32" | "i64" | "i128" | "isize" | "u8" | "u16" | "u32" | "u64" | "u128" | "usize" |
      "f16" | "f32" | "f64" | "f128" }
  | { readonly kind: "adt" | "closure" | "coroutine-closure" | "coroutine" | "coroutine-witness";
      readonly definition: RustNativeDefinitionId; readonly arguments: readonly RustNativeArgument[] }
  | { readonly kind: "foreign"; readonly definition: RustNativeDefinitionId }
  | { readonly kind: "array"; readonly element: number; readonly length: number }
  | { readonly kind: "pattern"; readonly base: number; readonly pattern: RustNativeTypePattern }
  | { readonly kind: "slice"; readonly element: number }
  | { readonly kind: "raw-pointer"; readonly pointee: number; readonly mutable: boolean }
  | { readonly kind: "reference"; readonly region: RustNativeRegion; readonly pointee: number; readonly mutable: boolean }
  | { readonly kind: "function"; readonly definition: RustNativeDefinitionId; readonly arguments: readonly RustNativeArgument[];
      readonly signature: RustNativeBinder<RustNativeSignature> }
  | { readonly kind: "function-pointer"; readonly signature: RustNativeBinder<RustNativeSignature> }
  | { readonly kind: "unsafe-binder"; readonly binder: RustNativeBinder<number> }
  | { readonly kind: "dynamic"; readonly predicates: readonly RustNativeBinder<RustNativeExistential>[]; readonly region: RustNativeRegion }
  | { readonly kind: "tuple"; readonly elements: readonly number[] }
  | { readonly kind: "alias"; readonly alias: RustNativeAlias & { readonly sort: "type" }; readonly rigid: boolean }
  | { readonly kind: "parameter"; readonly index: number; readonly name: string }
  | { readonly kind: "bound"; readonly binder: RustNativeBoundIndex; readonly variable: number; readonly declaration: RustNativeBoundType }
  | { readonly kind: "placeholder"; readonly universe: number; readonly variable: number; readonly declaration: RustNativeBoundType }
  | { readonly kind: "inference"; readonly category: "type" | "integer" | "float" | "fresh-type" | "fresh-integer" | "fresh-float"; readonly index: number };

export type RustNativeConstant =
  | { readonly kind: "parameter"; readonly index: number; readonly name: string }
  | { readonly kind: "bound"; readonly binder: RustNativeBoundIndex; readonly variable: number }
  | { readonly kind: "placeholder"; readonly universe: number; readonly variable: number }
  | { readonly kind: "inference"; readonly category: "constant" | "fresh-constant"; readonly index: number }
  | { readonly kind: "alias"; readonly alias: RustNativeAlias & { readonly sort: "constant" }; readonly rigid: boolean }
  | { readonly kind: "scalar"; readonly type: number; readonly bytes: number; readonly bits: string }
  | { readonly kind: "aggregate"; readonly type: number; readonly fields: readonly number[] }
  | { readonly kind: "expression"; readonly operation: RustNativeConstantOperation; readonly arguments: readonly RustNativeArgument[] };

export const rustNativeConstantOperations = [
  "add", "add-unchecked", "add-with-overflow", "sub", "sub-unchecked", "sub-with-overflow",
  "mul", "mul-unchecked", "mul-with-overflow", "div", "rem", "bit-xor", "bit-and", "bit-or",
  "shl", "shl-unchecked", "shr", "shr-unchecked", "eq", "lt", "le", "ne", "ge", "gt", "cmp", "offset",
  "not", "neg", "pointer-metadata", "call", "as", "use",
] as const;
export type RustNativeConstantOperation = typeof rustNativeConstantOperations[number];

export type RustNativeClause =
  | { readonly kind: "trait"; readonly definition: RustNativeDefinitionId; readonly arguments: readonly RustNativeArgument[];
      readonly polarity: "positive" | "negative" }
  | { readonly kind: "region-outlives"; readonly longer: RustNativeRegion; readonly shorter: RustNativeRegion }
  | { readonly kind: "type-outlives"; readonly type: number; readonly region: RustNativeRegion }
  | { readonly kind: "projection"; readonly alias: RustNativeAlias; readonly term: RustNativeArgument }
  | { readonly kind: "constant-type"; readonly constant: number; readonly type: number }
  | { readonly kind: "well-formed"; readonly term: RustNativeArgument }
  | { readonly kind: "constant-evaluatable"; readonly constant: number }
  | { readonly kind: "host-effect"; readonly definition: RustNativeDefinitionId; readonly arguments: readonly RustNativeArgument[];
      readonly constness: "const" | "maybe" }
  | { readonly kind: "unstable-feature"; readonly name: string };

export interface RustNativeGenerics {
  readonly parent: RustNativeDefinitionId | null;
  readonly parentCount: number;
  readonly hasSelf: boolean;
  readonly parameters: readonly RustNativeParameter[];
  readonly predicatesParent: RustNativeDefinitionId | null;
  readonly predicates: readonly RustNativeBinder<RustNativeClause>[];
}
export interface RustNativeParameter {
  readonly definition: RustNativeDefinitionId;
  readonly index: number;
  readonly name: string;
  readonly pureWrtDrop: boolean;
  readonly value:
    | { readonly kind: "lifetime" }
    | { readonly kind: "type"; readonly synthetic: boolean; readonly default: number | null }
    | { readonly kind: "constant"; readonly type: number; readonly default: number | null };
}

File: tools/rust-source-provider/src/type_model.rs
use serde::Serialize;

use crate::evidence::DefinitionId;

pub type TypeId = u32;
pub type ConstantId = u32;

#[derive(Serialize)]
pub struct TypeRow {
    pub id: TypeId,
    pub value: Type,
}

#[derive(Serialize)]
#[serde(tag = "kind", rename_all = "kebab-case", rename_all_fields = "camelCase")]
pub enum Type {
    Primitive { name: &'static str },
    Adt { definition: DefinitionId, arguments: Vec<Argument> },
    Foreign { definition: DefinitionId },
    Array { element: TypeId, length: ConstantId },
    Pattern { base: TypeId, pattern: Pattern },
    Slice { element: TypeId },
    RawPointer { pointee: TypeId, mutable: bool },
    Reference { region: Region, pointee: TypeId, mutable: bool },
    Function { definition: DefinitionId, arguments: Vec<Argument>, signature: Binder<Signature> },
    FunctionPointer { signature: Binder<Signature> },
    UnsafeBinder { binder: Binder<TypeId> },
    Dynamic { predicates: Vec<Binder<Existential>>, region: Region },
    Closure { definition: DefinitionId, arguments: Vec<Argument> },
    CoroutineClosure { definition: DefinitionId, arguments: Vec<Argument> },
    Coroutine { definition: DefinitionId, arguments: Vec<Argument> },
    CoroutineWitness { definition: DefinitionId, arguments: Vec<Argument> },
    Tuple { elements: Vec<TypeId> },
    Alias { alias: Alias, rigid: bool },
    Parameter { index: u32, name: String },
    Bound { binder: BoundIndex, variable: u32, declaration: BoundType },
    Placeholder { universe: u32, variable: u32, declaration: BoundType },
    Inference { category: &'static str, index: u32 },
}

#[derive(Serialize)]
#[serde(tag = "kind", rename_all = "kebab-case")]
pub enum Pattern {
    Range { start: ConstantId, end: ConstantId },
    Or { patterns: Vec<Pattern> },
    NotNull,
}

#[derive(Serialize)]
pub struct ConstantRow {
    pub id: ConstantId,
    pub value: Constant,
}

#[derive(Serialize)]
#[serde(tag = "kind", rename_all = "kebab-case")]
pub enum Constant {
    Parameter { index: u32, name: String },
    Bound { binder: BoundIndex, variable: u32 },
    Placeholder { universe: u32, variable: u32 },
    Inference { category: &'static str, index: u32 },
    Alias { alias: Alias, rigid: bool },
    Scalar { r#type: TypeId, bytes: u64, bits: String },
    Aggregate { r#type: TypeId, fields: Vec<ConstantId> },
    Expression { operation: &'static str, arguments: Vec<Argument> },
}

#[derive(Serialize)]
#[serde(tag = "kind", rename_all = "kebab-case")]
pub enum Argument {
    Type { id: TypeId },
    Constant { id: ConstantId },
    Lifetime { region: Region },
}

#[derive(Serialize)]
pub struct Alias {
    pub sort: &'static str,
    pub category: &'static str,
    pub definition: DefinitionId,
    pub arguments: Vec<Argument>,
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Signature {
    pub inputs: Vec<TypeId>,
    pub output: TypeId,
    pub variadic: bool,
    pub unsafe_call: bool,
    pub abi: &'static str,
}

#[derive(Serialize)]
pub struct Binder<Value> {
    pub variables: Vec<Variable>,
    pub value: Value,
}

#[derive(Serialize)]
#[serde(tag = "kind", rename_all = "kebab-case")]
pub enum Variable {
    Type { declaration: BoundType },
    Lifetime { declaration: BoundRegion },
    Constant,
}

#[derive(Serialize)]
#[serde(tag = "kind", rename_all = "kebab-case")]
pub enum BoundIndex {
    Bound { depth: u32 },
    Canonical,
}

#[derive(Serialize)]
#[serde(tag = "kind", rename_all = "kebab-case")]
pub enum BoundType {
    Anonymous,
    Named { definition: DefinitionId },
}

#[derive(Serialize)]
#[serde(tag = "kind", rename_all = "kebab-case")]
pub enum BoundRegion {
    Anonymous,
    Printed { name: String },
    Named { definition: DefinitionId },
    ClosureEnvironment,
}

#[derive(Serialize)]
#[serde(tag = "kind", rename_all = "kebab-case")]
pub enum LateRegion {
    Anonymous { index: u32 },
    Printed { index: u32, name: String },
    Named { definition: DefinitionId },
    ClosureEnvironment,
}

#[derive(Serialize)]
#[serde(tag = "kind", rename_all = "kebab-case")]
pub enum Region {
    Early { index: u32, name: String },
    Bound { binder: BoundIndex, variable: u32, declaration: BoundRegion },
    Late { scope: DefinitionId, declaration: LateRegion },
    Static,
    Inference { index: u32 },
    Placeholder { universe: u32, variable: u32, declaration: BoundRegion },
    Erased,
}

#[derive(Serialize)]
#[serde(tag = "kind", rename_all = "kebab-case")]
pub enum Existential {
    Trait { definition: DefinitionId, arguments: Vec<Argument> },
    Projection { definition: DefinitionId, arguments: Vec<Argument>, term: Argument },
    AutoTrait { definition: DefinitionId },
}

#[derive(Serialize)]
#[serde(tag = "kind", rename_all = "kebab-case")]
pub enum Clause {
    Trait { definition: DefinitionId, arguments: Vec<Argument>, polarity: &'static str },
    RegionOutlives { longer: Region, shorter: Region },
    TypeOutlives { r#type: TypeId, region: Region },
    Projection { alias: Alias, term: Argument },
    ConstantType { constant: ConstantId, r#type: TypeId },
    WellFormed { term: Argument },
    ConstantEvaluatable { constant: ConstantId },
    HostEffect { definition: DefinitionId, arguments: Vec<Argument>, constness: &'static str },
    UnstableFeature { name: String },
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Generics {
    pub parent: Option<DefinitionId>,
    pub parent_count: usize,
    pub has_self: bool,
    pub parameters: Vec<Parameter>,
    pub predicates_parent: Option<DefinitionId>,
    pub predicates: Vec<Binder<Clause>>,
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Parameter {
    pub definition: DefinitionId,
    pub index: u32,
    pub name: String,
    pub pure_wrt_drop: bool,
    pub value: ParameterKind,
}

#[derive(Serialize)]
#[serde(tag = "kind", rename_all = "kebab-case")]
pub enum ParameterKind {
    Lifetime,
    Type { synthetic: bool, default: Option<TypeId> },
    Constant { r#type: TypeId, default: Option<ConstantId> },
}
